File: katas/python/OOP/abstraction/abstraction_good.py
# Cumplimiento de Abstracción: Oculta detalles de implementación
# ✅ Solución: Interfaz simple que oculta la complejidad interna


class EmailSender:
    # ✅ Detalles internos son privados
    def __init__(self, username: str, password: str):
        self._smtp_server: str = "smtp.gmail.com"
        self._smtp_port: int = 587
        self._username: str = username
        self._password: str = password
        self._is_connected: bool = False

    # ✅ Método público simple que oculta toda la complejidad
    def send_email(self, to: str, subject: str, body: str) -> bool:
        try:
            # Internamente maneja todos los detalles
            self._connect()
            self._authenticate()
            message = self._build_message(to, subject, body)
            self._send(message)
            self._disconnect()

            print(f"Email enviado exitosamente a {to}")
            return True
        except Exception as error:
            print(f"Error al enviar email: {error}")
            return False

    # ✅ Métodos privados ocultan la implementación
    def _connect(self) -> None:
        print(f"Conectando a {self._smtp_server}:{self._smtp_port}...")
        self._is_connected = True

    def _authenticate(self) -> None:
        if not self._username or not self._password:
            raise Exception("Credenciales inválidas")
        print(f"Autenticando usuario {self._username}...")

    def _build_message(self, to: str, subject: str, body: str) -> str:
        return f"To: {to}\nSubject: {subject}\n\n{body}"

    def _send(self, message: str) -> None:
        if not self._is_connected:
            raise Exception("No conectado al servidor")
        print("Enviando mensaje...")

    def _disconnect(self) -> None:
        print("Desconectando del servidor...")
        self._is_connected = False


if __name__ == "__main__":
    # ✅ Usuario solo necesita saber la interfaz pública simple
    print("=== Cumplimiento de Abstracción ===")

    email_sender = EmailSender("usuario@gmail.com", "password123")

    # ✅ Un solo método simple - toda la complejidad está oculta
    email_sender.send_email(
        "destinatario@gmail.com", "Hola", "Este es el cuerpo del mensaje"
    )

    # ✅ Usuario no necesita conocer detalles internos como:
    # - Conexión al servidor
    # - Autenticación
    # - Construcción del mensaje
    # - Desconexión
    # Todo está abstraído en un método simple

File: katas/php/OOP/abstraction/abstraction-good.php
<?php
// Cumplimiento de Abstracción: Oculta detalles de implementación
// ✅ Solución: Interfaz simple que oculta la complejidad interna

class EmailSender {
    // ✅ Detalles internos son privados
    private string $smtpServer = "smtp.gmail.com";
    private int $smtpPort = 587;
    private string $username = "";
    private string $password = "";
    private bool $isConnected = false;

    // ✅ Constructor simplifica la configuración
    public function __construct(string $username, string $password) {
        $this->username = $username;
        $this->password = $password;
    }

    // ✅ Método público simple que oculta toda la complejidad
    public function sendEmail(string $to, string $subject, string $body): bool {
        try {
            // Internamente maneja todos los detalles
            $this->connect();
            $this->authenticate();
            $message = $this->buildMessage($to, $subject, $body);
            $this->send($message);
            $this->disconnect();

            echo "Email enviado exitosamente a $to" . PHP_EOL;
            return true;
        } catch (Exception $error) {
            echo "Error al enviar email: {$error->getMessage()}" . PHP_EOL;
            return false;
        }
    }

    // ✅ Métodos privados ocultan la implementación
    private function connect(): void {
        echo "Conectando a {$this->smtpServer}:{$this->smtpPort}..." . PHP_EOL;
        $this->isConnected = true;
    }

    private function authenticate(): void {
        if (!$this->username || !$this->password) {
            throw new Exception("Credenciales inválidas");
        }
        echo "Autenticando usuario {$this->username}..." . PHP_EOL;
    }

    private function buildMessage(string $to, string $subject, string $body): string {
        return "To: $to\nSubject: $subject\n\n$body";
    }

    private function send(string $message): void {
        if (!$this->isConnected) {
            throw new Exception("No conectado al servidor");
        }
        echo "Enviando mensaje..." . PHP_EOL;
    }

    private function disconnect(): void {
        echo "Desconectando del servidor..." . PHP_EOL;
        $this->isConnected = false;
    }
}

// ✅ Usuario solo necesita saber la interfaz pública simple
echo "=== Cumplimiento de Abstracción ===" . PHP_EOL;

$emailSender = new EmailSender("usuario@gmail.com", "password123");

// ✅ Un solo método simple - toda la complejidad está oculta
$emailSender->sendEmail(
    "destinatario@gmail.com",
    "Hola",
    "Este es el cuerpo del mensaje"
);

// ✅ Usuario no necesita conocer detalles internos como:
// - Conexión al servidor
// - Autenticación
// - Construcción del mensaje
// - Desconexión
// Todo está abstraído en un método simple
